core/index/index_writer.hpp

```diff
@@ -48,131 +50,204 @@
 namespace irs {
 
 class Comparer;
-class bitvector;
 struct directory;
-class directory_reader;
-
-class readers_cache final : util::noncopyable {
- public:
-  struct key_t {
-    // cppcheck-suppress noExplicitConstructor
-    key_t(const segment_meta& meta);  // implicit constructor
-
-    bool operator==(const key_t& other) const noexcept {
-      return name == other.name && version == other.version;
-    }
-
-    std::string name;
-    uint64_t version;
-  };
-
-  struct key_hash_t {
-    size_t operator()(const key_t& key) const noexcept {
-      return hash_utils::Hash(key.name);
-    }
-  };
-
-  // cppcheck-suppress constParameter
-  explicit readers_cache(directory& dir) noexcept : dir_(dir) {}
-
-  void clear() noexcept;
-  segment_reader emplace(const segment_meta& meta);
-  size_t purge(const absl::flat_hash_set<key_t, key_hash_t>& segments) noexcept;
-
- private:
-  std::mutex lock_;
-  absl::flat_hash_map<key_t, segment_reader, key_hash_t> cache_;
-  directory& dir_;
-};  // readers_cache
 
 // Defines how index writer should be opened
 enum OpenMode {
   // Creates new index repository. In case if repository already
   // exists, all contents will be cleared.
   OM_CREATE = 1,
 
-  // Opens existsing index repository. In case if repository does not
+  // Opens existing index repository. In case if repository does not
   // exists, error will be generated.
   OM_APPEND = 2,
 };
 
 ENABLE_BITMASK_ENUM(OpenMode);
 
+// A set of candidates denoting an instance of consolidation
+using Consolidation = std::vector<const SubReader*>;
+using ConsolidationView = std::span<const SubReader* const>;
+
+struct CandidateHash {
+  size_t operator()(const SubReader* segment) const noexcept {
+    return absl::HashOf(segment->Meta().name);
+  }
+};
+
+struct CandidateEqual {
+  bool operator()(const SubReader* lhs, const SubReader* rhs) const noexcept {
+    return lhs->Meta().name == rhs->Meta().name;
+  }
+};
+
+// segments that are under consolidation
+using ConsolidatingSegments =
+  absl::flat_hash_set<const SubReader*, CandidateHash, CandidateEqual>;
+
+// Mark consolidation candidate segments matching the current policy
+// candidates the segments that should be consolidated
+// in: segment candidates that may be considered by this policy
+// out: actual segments selected by the current policy
+// dir the segment directory
+// meta the index meta containing segments to be considered
+// Consolidating_segments segments that are currently in progress
+// of consolidation
+// Final candidates are all segments selected by at least some policy
+using ConsolidationPolicy =
+  std::function<void(Consolidation& candidates, const IndexReader& index,
+                     const ConsolidatingSegments& consolidating_segments)>;
+
+enum class ConsolidationError : uint32_t {
```

Comment:
> uint32_t

Why?

Reply:
Why not? (c) ;-)

Reply:
If there is no propose I suggest to remove it or write uint8_t